### PR TITLE
Week4

### PR DESCRIPTION
**Pull request by hepnerthomas**
# Week 4: Hardhat. Unit Testing, Mutation testing, and Static Analysis

## Learning Objectives
- Learn how to set up and use the hardhat environment
- Learn how to test every possible state change in Ethereum
- Learn how to fork the mainnet and test against it
- Learn how and why to use solhint, eslint, and prettier and why those tools are important
- Learn and use static analysis and mutation testing

### Old cleanup

- [x] Clean up your old code with solhint and prettier. You might need to adjust the maximum line length if the formatting looks funny.
- [x] Run slither on the final version of your previous assignments. Did it find any errors or were they false positives?

Week1 Assignment: Running Slither displayed this output:
Slither

Slither detected 17 results. None of them were false positives, although all but the two reentrancy issues were either low or informational severerity issues.

Slither caught these issues which I corrected:
- Tokens.buyTokens().totalSupply (contracts/Token.sol) shadows
- Token.buyTokens().decimals (contracts/Token.sol#166) shadows
- Token.sellBack(uint256).decimals (contracts/Token.sol#193) shadows
- Different versions of Solidity are used (2 issues)
- Token.mintPrice (contracts/Token.sol#16) should be constant
- Token.refundPrice (contracts/Token.sol#17) should be constant
- Token.owner (contracts/Token.sol#19) should be immutable
- solc-0.8.17 is not recommended for deployment
- Reentrancy in Token.sellBack(uint256)
- Reentrancy in Token.withdrawETH()

Week 2 Assignment: Running Slither displayed this [output](slitherOutputWeek3.txt):

Slither detected 30 results. Here's the ones I corrected:

- ForgeMetals.COOLDOWN_PERIOD_SECONDS (contracts/ForgeMetals.sol#40) should be constant
- ForgeMetals.metals (contracts/ForgeMetals.sol#38) should be immutable
- Different versions of Solidity are used: Version used: ['^0.8.0', '^0.8.1', '^0.8.9']
- ForgeMetals.tradeMetal(uint256,uint256) (contracts/ForgeMetals.sol#113-132) contains a tautology or contradiction
- ForgeMetals.mineMetal(uint256) (contracts/ForgeMetals.sol#54-66) uses a dangerous strict equality
- Reentrancy in ForgeMetals.mineMetal(uint256) (contracts/ForgeMetals.sol#54-66)

Errors I ignored and added to Slither Triage file:
I intentionally ignored all of the Informational positives related to OpenZeppelin. I ignored a Dangerous Comparison related to block.timestamp as this is critical to how the ForgeMetals contract works. 

### Add unit tests to the Partial Refund assignment from week 1 and add unit tests to the forging dapp from week 3

- [x] One of the assignments should have 100% line and branch coverage. The other one can have 90% for both.
- [x] Run mutation testing on the assignment that has 100% coverage. Did you discover any faulty tests?

Yes, and week1 assignment had unnecessary branching in my Solidity. See the `week1/mutation-testing` folder. For instance, I used an `else if` when an `else` would have been sufficient for code functionality. The tests were not faulty, but they were insufficient at capturing all of the cases in my branching logic, especially for
seeing the expected behavior for when two variables were equal to each other in 'less than or equals' or 'greater than or equals' conditions.

I did not have any tests for my week3 assignment, but adding tests caught a || that should have been a && in the tradeMetal of the week3 forging assignment.

Ethernaut

- https://ethernaut.openzeppelin.com/level/0xBd886a37faD1f596221f33ca568122815ED48c81

- https://ethernaut.openzeppelin.com/level/0x6F9cf195B9B4c1259E8FCe5b4e30F7142f779DeA

- https://ethernaut.openzeppelin.com/level/0x40F5513a90fb7e2ac2C3E12A6d16B9279D1e94Ed

## Description of Deliverables

- [x] Show the github commit that fixes the old code
- [x] Do a write-up of errors you discovered. If none, pick 3 false positives and describe why they are false positives.
- [x] Show the github commit that adds the unit tests.

### CODE REVIEW

File: week1/test/Token.js [hunk 1]
```diff
@@ -305,18 +350,76 @@ describe("Token contract", function () {
...
+          value: parseEther("1.0"),
+        });
+        const amount = 1000;
+        const value = BigNumber.from(parseUnits(String(amount), decimals));
+
+        // addr1 should have 1000 tokens
+        expect(await hardhatToken.balanceOf(addr1.address)).to.equal(value);
+
+        // should be 1 ether in the contract
+        expect(await await ethers.provider.getBalance(hardhatToken.address)).to.equal(parseEther("1.0"));
```
Comment (dhruvinparikh) [inline week1/test/Token.js:420]: - [ ] assert expected  eth balance for `addr`

File: week1/test/Token.js [hunk 2]
```diff
@@ -412,6 +515,28 @@ describe("Token contract", function () {
...
 
+      it("Should revert if msg.sender balance is less than the amount", async function () {
+        const { hardhatToken, owner, addr1, addr2 } = await loadFixture(deployTokenFixture);
+        const senderBalance = 1999;
+        const senderBalanceFormatted = BigNumber.from(parseUnits(String(senderBalance), decimals));
+        await hardhatToken.changeBalanceAtAddress(addr1.address, senderBalanceFormatted);
+
+        const tokensToRefund = 2000;
+        const tokensToRefundFormatted = BigNumber.from(parseUnits(String(tokensToRefund), decimals));
+        await expect(hardhatToken.connect(addr1).sellBack(tokensToRefundFormatted)).to.be.revertedWith("Not enough tokens");
```
Comment (dhruvinparikh) [inline week1/test/Token.js:526]: - [ ] assert the eth balance of seller
- [ ] assert token balance of seller
- [ ] assert eth balance of token sale contract

File: week1/test/Token.js [hunk 3]
```diff
@@ -412,6 +515,28 @@ describe("Token contract", function () {
...
+        const { hardhatToken, owner, addr1, addr2 } = await loadFixture(deployTokenFixture);
+        const senderBalance = 1999;
+        const senderBalanceFormatted = BigNumber.from(parseUnits(String(senderBalance), decimals));
+        await hardhatToken.changeBalanceAtAddress(addr1.address, senderBalanceFormatted);
+
+        const tokensToRefund = 0;
+        const tokensToRefundFormatted = BigNumber.from(parseUnits(String(tokensToRefund), decimals));
+        await expect(hardhatToken.connect(addr1).sellBack(tokensToRefundFormatted)).to.be.revertedWith("Must send tokens to receive ETH.");
+      });
+
```
Comment (dhruvinparikh) [inline week1/test/Token.js:539]: - [ ] Also assert ETH balance in "Withdraw ETH" case.

File: week3/test/ForgeMetals.js [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+      await time.increase(60);
+      await forgeMetalsContract.connect(addr1).mineMetal(1);
+      await forgeMetalsContract.connect(addr1).forgeAlloy(0, 1);
+
+      const balanceMetalId1 = await forgeMetalsContract.getBalance(addr1.address, 0);
+      const balanceMetalId2 = await forgeMetalsContract.getBalance(addr1.address, 1);
+      const balance = await forgeMetalsContract.getBalance(addr1.address, 3);
+
+      await expect(balanceMetalId1).to.equal(0);
+      await expect(balanceMetalId2).to.equal(0);
```
Comment (dhruvinparikh) [inline week3/test/ForgeMetals.js:138]:
```suggestion
      await expect(balanceMetalId1).to.equal(oldBalanceMetalId1 - 1);
      await expect(balanceMetalId2).to.equal(oldBalanceMetalId2 - 1);
```

File: week3/test/ForgeMetals.js [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+describe("Forge Metals contract", function () {
+  async function deployForgeMetalsFixture() {
+    const ForgeMetals = await ethers.getContractFactory("ForgeMetals");
+    const [owner, addr1, addr2] = await ethers.getSigners();
+    const forgeMetalsContract = await ForgeMetals.deploy();
+    await forgeMetalsContract.deployed();
+    return { ForgeMetals, forgeMetalsContract, owner, addr1, addr2 };
+  }
+
```
Comment (dhruvinparikh) [inline week3/test/ForgeMetals.js:15]: you can load fixture in `beforeEach` hook

**Review by dhruvinparikh — CHANGES_REQUESTED** [review]

File: week1/test/Token.js (hunk 1 above)
Comment (hepnerthomas) [reply week1/test/Token.js:420]: done

File: week1/test/Token.js (hunk 2 above)
Comment (hepnerthomas) [reply week1/test/Token.js:526]: done

File: week1/test/Token.js (hunk 3 above)
Comment (hepnerthomas) [reply week1/test/Token.js:539]: done

File: week3/test/ForgeMetals.js (hunk 4 above)
Comment (hepnerthomas) [reply week3/test/ForgeMetals.js:138]: done

File: week3/test/ForgeMetals.js (hunk 5 above)
Comment (hepnerthomas) [reply week3/test/ForgeMetals.js:23]: done

**Pull request merged**